Objective-C/CBLReadOnlyDocument.mm

```diff
@@ -86,53 +92,72 @@ - (C4Database*) c4db {
 
 
 - (void) setC4Doc: (CBLC4Document*)c4doc {
-    _c4Doc = c4doc;
-
-    if (c4doc) {
-        FLDict root = nullptr;
-        C4Slice body = c4doc.selectedRev.body;
-        if (body.size > 0)
-            root = FLValue_AsDict(FLValue_FromTrustedData({body.buf, body.size}));
-        self.data = [[CBLFLDict alloc] initWithDict: root datasource: c4doc database: _database];
-    } else {
-        self.data = nil;
+    CBL_LOCK(self.lock) {
+        _c4Doc = c4doc;
+        
+        if (c4doc) {
+            FLDict root = nullptr;
+            C4Slice body = c4doc.selectedRev.body;
+            if (body.size > 0)
+                root = FLValue_AsDict(FLValue_FromTrustedData({body.buf, body.size}));
+            self.data = [[CBLFLDict alloc] initWithDict: root datasource: c4doc database: _database];
+        } else {
+            self.data = nil;
+        }
+    }
+}
+
+
+- (CBLC4Document*) c4Doc {
+    CBL_LOCK(self.lock) {
+        return _c4Doc;
     }
 }
 
 
 - (BOOL) selectConflictingRevision {
-    if (!_c4Doc || !c4doc_selectNextLeafRevision(_c4Doc.rawDoc, false, true, nullptr))
-        return NO;
-    self.c4Doc = _c4Doc;     // This will update to the selected revision
-    return YES;
+    CBL_LOCK(self.lock) {
+        if (!_c4Doc || !c4doc_selectNextLeafRevision(_c4Doc.rawDoc, false, true, nullptr))
+            return NO;
+        self.c4Doc = _c4Doc;     // This will update to the selected revision
+        return YES;
+    }
 }
 
 
 - (BOOL) selectCommonAncestorOfDoc: (CBLReadOnlyDocument*)doc1
                             andDoc: (CBLReadOnlyDocument*)doc2
 {
-    CBLStringBytes rev1(doc1.revID), rev2(doc2.revID);
-    if (!_c4Doc || !c4doc_selectCommonAncestorRevision(_c4Doc.rawDoc, rev1, rev2)
-                || !c4doc_hasRevisionBody(_c4Doc.rawDoc))
-        return NO;
-    self.c4Doc = _c4Doc;     // This will update to the selected revision
-    return YES;
+    CBL_LOCK(self.lock) {
+        CBLStringBytes rev1(doc1.revID), rev2(doc2.revID);
+        if (!_c4Doc || !c4doc_selectCommonAncestorRevision(_c4Doc.rawDoc, rev1, rev2)
```

Comment:
Yes, this could impact CBLDocument, especially _dict because it will get out of sync. If CBLDocument uses the same lock here to lock accessing to _dict (self.dict), I think the consistency state should be protected.